src/app/admin/page.tsx

```diff
@@ -1,60 +1,251 @@
       try {
         const fetchedUser = await getUser();
-        setUser(fetchedUser);
+        setState((prevState) => ({
+          ...prevState,
+          user: fetchedUser,
+        }));
       } catch (error) {
         console.error("Error fetching user:", error);
+        router.push("/");
       } finally {
-        setIsLoading(false);
+        setIsLoadingUser(false);
       }
     };
 
     fetchUser();
-  }, []);
+  }, [router]);
+
+  const { users, handleRoleChange } = useUserManagement(state.user);
+
+  const openModal = (user: User) => {
+    setSelectedUser(user);
+    setIsModalOpen(true);
+  };
+
+  const closeModal = () => {
+    setIsModalOpen(false);
+    setSelectedUser(null);
+  };
 
-  if (isLoading) {
-    return <div className="flex min-h-screen items-center justify-center text-3xl">Loading...</div>;
+  if (isLoadingUser) {
+    return (
+      <div className="flex min-h-screen items-center justify-center text-3xl">
+        Loading...
+      </div>
+    );
   }
 
-  if (!user || !user.admin) {
+  if (!state.user || state.user.access === "user") {
     router.push("/");
-  } else {
-    return (
-      <div>
-        <Navbar />
-
-        <div className="mx-auto mt-12 flex max-w-6xl flex-col px-8 pb-8 ">
-          <div className="mb-6 flex flex-col gap-1">
-            <h1 className="text-balance text-left text-5xl font-extrabold lg:text-6xl">
-              Admin dashboard
-            </h1>
-            <p className="w-full text-pretty text-lg opacity-70 lg:text-xl">
-              Click the subject you wish to add articles to
-            </p>
-          </div>
-          <div className="grid grid-cols-3 gap-4 lg:gap-8">
-            <APLibrary />
+    return null;
+  }
+
+  return (
+    <div>
+      <Navbar />
+
+      <div className="mx-auto mt-12 flex max-w-6xl flex-col px-8 pb-8 ">
+        <div className="mb-6 flex flex-col gap-1">
+          <h1 className="text-balance text-left text-5xl font-extrabold lg:text-6xl">
+            Admin dashboard
+          </h1>
+
+          {state.user.access === "admin" && (
+            <>
+              <p className="w-full text-pretty text-lg opacity-70 sm:text-lg lg:text-xl">
+                Change the role of a user
+              </p>
+              <div className="my-4 rounded-2xl border-2 p-4">
+                <div className="rounded-md bg-white p-4">
+                  <input
+                    type="text"
+                    className="mb-3 w-full rounded-md border p-2"
+                    placeholder="Search for a user..."
+                    value={state.searchTermUsers}
+                    onChange={(e) =>
+                      setState((prevState) => ({
+                        ...prevState,
+                        searchTermUsers: e.target.value,
+                      }))
+                    }
+                  />
+                  <ul className="class-list max-h-60 overflow-y-auto">
+                    {users!.map((u) => (
```

Comment:
questionable usage of non-null coercion, leave a comment explaining how and why `users` will never be nullable